crates/cargo-test-macro/src/lib.rs

```diff
@@ -144,18 +147,61 @@ pub fn cargo_test(attr: TokenStream, item: TokenStream) -> TokenStream {
                     continue;
                 }
             }
+            TokenTree::Ident(i) => {
+                // The first time through it will be `fn` the second time is the
+                // name of the test.
+                if test_name.is_none() && num == 1 {
+                    test_name = Some(i.to_string())
+                } else {
+                    num += 1;
+                }
+                ret.extend(Some(TokenTree::Ident(i)));
+                continue;
+            }
             other => {
                 ret.extend(Some(other));
                 continue;
             }
         };
 
-        let mut new_body = to_token_stream(
-            r#"let _test_guard = {
+        let name = &test_name
+            .clone()
+            .map(|n| n.split("::").next().unwrap().to_string())
+            .unwrap();
+        let mut new_body = to_token_stream(&format!(
+            r#"let _test_guard = {{
                 let tmp_dir = option_env!("CARGO_TARGET_TMPDIR");
-                cargo_test_support::paths::init_root(tmp_dir)
-            };"#,
-        );
+                // looks at `file!()` to figure out where the test's directory should be.
+                // test fn `workspaces::workspace_in_git`
+                //   -> `file!` expands to `tests/testsuite/workspaces.rs`
+                //   -> `testsuite/workspaces`
+                // test fn `cargo_remove::cargo_remove::invalid_arg::case`
+                //   -> `file!` expands to `tests/testsuite/cargo_remove/invalid_arg.rs`
+                //   -> `testsuite/cargo_remove/invalid_arg`
+                // test fn `build-std::cross_custom`
+                //   -> `file!` expands to `tests/build-std/main.rs`
+                //   -> `build-std/main`
+                let test_dir: std::path::PathBuf = std::path::PathBuf::from(std::file!())
+                    .components()
+                    // skips `tests`
+                    .skip(1)
+                    // gets 0-3 of the nest componests. 3 is needed since we use
+                    // the file name for `snapbox` tests as the test name.
+                    .take(3)
+                    // remove `.rs` when using file name as test name
+                    .map(|c| c.as_os_str().to_str().unwrap().trim_end_matches(".rs"))
+                    .collect();
+                // some `snapbox` tests (i.e. `cargo_remove` ) use `case` as the test name.
+                // In this case we should use the file name as the test name. If it doesn't
+                // `case` add `name` to the `Pathbuf`
+                let test_dir = if "{name}" == "case" {{
+                    test_dir
```

Comment:
Is there a reason to do all of this rather than treating this more naively?  For snapbox tests, we'd end up just having the tmp dir be one level deeper.  As long as we aren't hitting length issues on Windows, that shouldn't be too much of a problem.

By skipping all of this / making the code more naively, we make it less brittle.  If someone comes along and tries to change anything about what we do, I fear they will get surprising results that will be hard to narrow down back to there.